util/env_windows.cc

```diff
@@ -0,0 +1,920 @@
+      NULL
+    );
+    if (h == INVALID_HANDLE_VALUE) {
+      *result = NULL;
+      return WindowsError(fname, GetLastError());
+    } else {
+      // get file descriptor
+      int fd = _open_osfhandle((intptr_t) h, 0);
+      if (fd == - 1) {
+        CloseHandle(h);
+        *result = NULL;
+        return WindowsError(fname, GetLastError());
+      }
+
+      // get FILE*
+      FILE* f = _fdopen(fd, "wb+");
+      if (f == NULL) {
+        _close(fd);
+        *result = NULL;
+        return WindowsError(fname, GetLastError());
+      }
+      *result = new FileLogger(f, &WindowsEnv::gettid);
+      return Status::OK();
+    }
+  }
+
+  virtual uint64_t NowMicros() {
+    return static_cast<uint64_t>(std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::high_resolution_clock::now().time_since_epoch()).count());
+  }
+
+  virtual void SleepForMicroseconds(int micros) {
+    std::this_thread::sleep_for(std::chrono::microseconds(micros));
+  }
+
+private:
+
+  // BGThread() is the body of the background thread
+  void BGThread();
+  static DWORD __stdcall BGThreadWrapper(void* arg) {
+    reinterpret_cast<WindowsEnv*>(arg)->BGThread();
+    return 0;
+  }
+
+  port::Mutex mu_;
+  port::CondVar bgsignal_;
+  std::unique_ptr<std::thread> bgthread_;
+  std::atomic<bool> started_bgthread_;
+
+  // Entry per Schedule() call
+  struct BGItem { void* arg; void (*function)(void*); };
+  typedef std::deque<BGItem> BGQueue;
+  BGQueue queue_;
+
+  WindowsLockTable locks_;
+  Limiter mmap_limit_;
+  Limiter fd_limit_;
+};
+
+// Return the maximum number of concurrent mmaps.
+static int MaxMmaps() {
+  if (mmap_limit >= 0) {
+    return mmap_limit;
+  }
+  // Up to 1000 mmaps for 64-bit binaries; none for smaller pointer sizes.
+  mmap_limit = sizeof(void*) >= 8 ? 1000 : 0;
+  return mmap_limit;
+}
+
+// Return the maximum number of read-only files to keep open.
+static intptr_t MaxOpenFiles() {
+  if (open_read_only_file_limit >= 0) {
+    return open_read_only_file_limit;
+  }
+
+  // Allow use of 20% of available file descriptors for read-only files.
+  open_read_only_file_limit = _getmaxstdio() / 5;
+
+  return open_read_only_file_limit;
+}
+
+WindowsEnv::WindowsEnv()
+    : bgsignal_(&mu_),
+      started_bgthread_(false),
+      mmap_limit_(MaxMmaps()),
+      fd_limit_(MaxOpenFiles()) {
+}
+
+void WindowsEnv::Schedule(void (*function)(void*), void* arg) {
+  MutexLock l(&mu_);
+
+  // Start background thread if necessary
+  if (!started_bgthread_) {
+    started_bgthread_ = true;
+    bgthread_.reset(new std::thread(std::bind(&WindowsEnv::BGThreadWrapper, this)));
+  }
+
+  // If the queue is currently empty, the background thread may currently be
+  // waiting.
+  if (queue_.empty()) {
+    bgsignal_.Signal();
+  }
+
+  // Add to priority queue
+  queue_.push_back(BGItem());
+  queue_.back().function = function;
+  queue_.back().arg = arg;
+}
+
+void WindowsEnv::BGThread() {
+  while (started_bgthread_) {
```

Comment:
This is accessing started_bgthread_ without holding a lock, and it is written by another thread.

The complication of termination should go away once you clean up the destructor above so it crashes without trying to shutdown this thread.